File: src/VisualBasic/CodeCracker/Usage/UnusedParametersAnalyzer.vb
Imports System.Collections.Immutable
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Usage
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class UnusedParametersAnalyzer
        Inherits DiagnosticAnalyzer

        Friend Const Title = "Unused parameters."
        Friend Const Message = "Parameter '{0}' is not used."
        Private Const Description = "When a method declares a parameter and does not use it might bring incorrect concolusions for anyone reading the code and anso demands the parameter when the method is called unnecessarily.
You should delete the parameter in such cases."

        Friend Shared Rule As New DiagnosticDescriptor(
        DiagnosticId.UnusedParameters.ToDiagnosticId(),
        Title,
        Message,
        SupportedCategories.Usage,
        DiagnosticSeverity.Warning,
        True,
        Description,
        HelpLink.ForDiagnostic(DiagnosticId.UnusedParameters),
        WellKnownDiagnosticTags.Unnecessary)

        Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
            Get
                Return ImmutableArray.Create(Rule)
            End Get
        End Property

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(AddressOf Analyzer, SyntaxKind.SubBlock, SyntaxKind.ConstructorBlock, SyntaxKind.FunctionBlock)
        End Sub

        Private Sub Analyzer(context As SyntaxNodeAnalysisContext)
            If (context.IsGenerated()) Then Return
            Dim methodOrConstructor = TryCast(context.Node, MethodBlockBaseSyntax)
            If methodOrConstructor Is Nothing Then Exit Sub
            Dim model = context.SemanticModel
            If Not IsCandidateForRemoval(methodOrConstructor, model) Then Exit Sub
            Dim parameters = methodOrConstructor.BlockStatement.ParameterList.Parameters.ToDictionary(Function(p) p, Function(p) model.GetDeclaredSymbol(p))
            Dim ctor = TryCast(methodOrConstructor, ConstructorBlockSyntax)
            ' TODO: Check if used in MyBase
            If methodOrConstructor.Statements.Any() Then
                Dim dataFlowAnalysis = model.AnalyzeDataFlow(methodOrConstructor.Statements.First, methodOrConstructor.Statements.Last)
                If Not dataFlowAnalysis.Succeeded Then Exit Sub
                For Each parameter In parameters
                    Dim parameterSymbol = parameter.Value
                    If parameterSymbol Is Nothing Then Continue For
                    If Not dataFlowAnalysis.ReadInside.Contains(parameterSymbol) AndAlso
                        Not dataFlowAnalysis.WrittenInside.Contains(parameterSymbol) Then
                        context = CreateDiagnostic(context, parameter.Key)
                    End If
                Next
            Else
                For Each parameter In parameters.Keys
                    context = CreateDiagnostic(context, parameter)
                Next
            End If
        End Sub

        Private Shared Function IsCandidateForRemoval(methodOrConstructor As MethodBlockBaseSyntax, semanticModel As SemanticModel) As Boolean
            If methodOrConstructor.BlockStatement.Modifiers.Any(Function(m) m.ValueText = "Partial" OrElse m.ValueText = "Overrides") OrElse
            Not methodOrConstructor.BlockStatement.ParameterList?.Parameters.Any() Then Return False

            Dim method = TryCast(methodOrConstructor, MethodBlockSyntax)
            If method IsNot Nothing Then
                If method.SubOrFunctionStatement.ImplementsClause IsNot Nothing Then Return False
                Dim methodSymbol = semanticModel.GetDeclaredSymbol(method)
                If methodSymbol Is Nothing Then Return False
                Dim typeSymbol = methodSymbol.ContainingType
                If typeSymbol.Interfaces.SelectMany(Function(i) i.GetMembers()).
                Any(Function(member) methodSymbol.Equals(typeSymbol.FindImplementationForInterfaceMember(member))) Then Return False

                If IsEventHandlerLike(method, semanticModel) Then Return False
            Else
                Dim constructor = TryCast(methodOrConstructor, ConstructorBlockSyntax)
                If constructor IsNot Nothing Then
                    If IsSerializationConstructor(constructor, semanticModel) Then Return False
                Else
                    Return False
                End If
            End If
            Return True
        End Function

        Private Shared Function IsSerializationConstructor(constructor As ConstructorBlockSyntax, model As SemanticModel) As Boolean
            If constructor.SubNewStatement.ParameterList.Parameters.Count <> 2 Then Return False
            Dim constructorSymbol = model.GetDeclaredSymbol(constructor)
            Dim typeSymbol = constructorSymbol?.ContainingType
            If If(Not typeSymbol?.Interfaces.Any(Function(i) i.ToString() = "System.Runtime.Serialization.ISerializable"), True) Then Return False
            If Not typeSymbol.GetAttributes().Any(Function(a) a.AttributeClass.ToString() = "System.SerializableAttribute") Then Return False
            Dim serializationInfoType = TryCast(model.GetTypeInfo(constructor.SubNewStatement.ParameterList.Parameters(0).AsClause.Type).Type, INamedTypeSymbol)
            If serializationInfoType Is Nothing Then Return False
            If Not serializationInfoType.AllBaseTypesAndSelf().Any(Function(type) type.ToString() = "System.Runtime.Serialization.SerializationInfo") Then Return False

            Dim streamContextType = TryCast(model.GetTypeInfo(constructor.SubNewStatement.ParameterList.Parameters(1).AsClause.Type).Type, INamedTypeSymbol)
            If streamContextType Is Nothing Then Return False
            Return streamContextType.AllBaseTypesAndSelf().Any(Function(type) type.ToString() = "System.Runtime.Serialization.StreamingContext")
        End Function

        Private Shared Function IsEventHandlerLike(method As MethodBlockSyntax, model As SemanticModel) As Boolean

            If method.SubOrFunctionStatement.ParameterList.Parameters.Count <> 2 OrElse
            method.IsKind(SyntaxKind.FunctionBlock) Then Return False

            Dim senderType = model.GetTypeInfo(method.SubOrFunctionStatement.ParameterList.Parameters(0).AsClause.Type).Type
            If senderType.SpecialType <> SpecialType.System_Object Then Return False
            Dim eventArgsType = TryCast(model.GetTypeInfo(method.SubOrFunctionStatement.ParameterList.Parameters(1).AsClause.Type).Type, INamedTypeSymbol)
            If eventArgsType Is Nothing Then Return False
            Return eventArgsType.AllBaseTypesAndSelf().Any(Function(type) type.ToString() = "System.EventArgs")
        End Function

        Private Function CreateDiagnostic(context As SyntaxNodeAnalysisContext, parameter As ParameterSyntax) As SyntaxNodeAnalysisContext
            Dim propsDic = New Dictionary(Of String, String)
            propsDic.Add("identifier", parameter.Identifier.Identifier.Text)
            Dim props = propsDic.ToImmutableDictionary()
            Dim diag = Diagnostic.Create(Rule, parameter.GetLocation(), props, parameter.Identifier.Identifier.ValueText)
            context.ReportDiagnostic(diag)
            Return context
        End Function
    End Class
End Namespace

File: src/CSharp/CodeCracker/Usage/UnusedParametersAnalyzer.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

namespace CodeCracker.CSharp.Usage
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class UnusedParametersAnalyzer : DiagnosticAnalyzer
    {
        internal const string Title = "Unused parameters";
        internal const string Message = "Parameter '{0}' is not used.";
        internal const string Category = SupportedCategories.Usage;
        const string Description = "A method with an unused parameter creates unnecessary confusion and should be deleted.";

        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
            DiagnosticId.UnusedParameters.ToDiagnosticId(),
            Title,
            Message,
            Category,
            DiagnosticSeverity.Warning,
            isEnabledByDefault: true,
            description: Description,
            customTags: WellKnownDiagnosticTags.Unnecessary,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.UnusedParameters));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);

        public override void Initialize(AnalysisContext context) =>
            context.RegisterSyntaxNodeAction(Analyzer, SyntaxKind.MethodDeclaration, SyntaxKind.ConstructorDeclaration);

        private static void Analyzer(SyntaxNodeAnalysisContext context)
        {
            if (context.IsGenerated()) return;
            var methodOrConstructor = context.Node as BaseMethodDeclarationSyntax;
            if (methodOrConstructor == null) return;
            var semanticModel = context.SemanticModel;
            if (!IsCandidateForRemoval(methodOrConstructor, semanticModel)) return;
            var parameters = methodOrConstructor.ParameterList.Parameters.ToDictionary(p => p, p => semanticModel.GetDeclaredSymbol(p));
            var ctor = methodOrConstructor as ConstructorDeclarationSyntax;

            if (ctor?.Initializer != null)
            {
                var symbolsTouched = new List<ISymbol>();
                foreach (var arg in ctor.Initializer.ArgumentList.Arguments)
                {
                    var dataFlowAnalysis = semanticModel.AnalyzeDataFlow(arg.Expression);
                    if (!dataFlowAnalysis.Succeeded) continue;
                    symbolsTouched.AddRange(dataFlowAnalysis.ReadInside);
                    symbolsTouched.AddRange(dataFlowAnalysis.WrittenInside);
                }
                var parametersToRemove = parameters.Where(p => symbolsTouched.Contains(p.Value)).ToList();
                foreach (var parameter in parametersToRemove)
                    parameters.Remove(parameter.Key);
            }

            var method = methodOrConstructor as MethodDeclarationSyntax;

            // It is legit for virtual methods to have parameters that aren't used in the default
            // base class implementation, but are only provided for sub classes instead.
            // See https://github.com/code-cracker/code-cracker/issues/872
            if (method?.Modifiers.Any(SyntaxKind.VirtualKeyword) == true) return;

            IEnumerable<SyntaxNode> methodChildren = methodOrConstructor.Body?.Statements;
            var expressionBody = (methodOrConstructor as MethodDeclarationSyntax)?.ExpressionBody;
            if (methodChildren == null && expressionBody != null)
                methodChildren = new[] { expressionBody };

            if (methodChildren?.Any() ?? false)
            {
                var identifiers = methodChildren
                    .SelectMany(s => s.DescendantNodesAndSelf())
                    .OfType<IdentifierNameSyntax>()
                    .ToList();
                foreach (var parameter in parameters)
                {
                    var used = identifiers
                        .Any(iName => IdentifierRefersToParam(iName, parameter.Key));

                    if (!used)
                    {
                        ReportDiagnostic(context, parameter.Key);
                    }
                }
                //
                // THIS IS THE RIGHT WAY TO DO THIS VERIFICATION.
                // BUT, WE HAVE TO WAIT FOR A "BUGFIX" FROM ROSLYN TEAM
                // IN DataFlowAnalysis
                //
                // https://github.com/dotnet/roslyn/issues/6967
                //
                //var dataFlowAnalysis = semanticModel.AnalyzeDataFlow(methodOrConstructor.Body);
                //if (!dataFlowAnalysis.Succeeded) return;
                //foreach (var parameter in parameters)
                //{

                //    var parameterSymbol = parameter.Value;
                //    if (parameterSymbol == null) continue;
                //    if (!dataFlowAnalysis.ReadInside.Contains(parameterSymbol) &&
                //        !dataFlowAnalysis.WrittenInside.Contains(parameterSymbol))
                //    {
                //        ReportDiagnostic(context, parameter.Key);
                //    }
                //}
            }
            else
            {
                foreach (var parameter in parameters.Keys)
                    ReportDiagnostic(context, parameter);
            }
        }

        private static bool IdentifierRefersToParam(IdentifierNameSyntax iName, ParameterSyntax param)
        {
            var identifierName = iName.Identifier.ToString();
            if (!identifierName.StartsWith("@")) identifierName = $"@{identifierName}";
            var parameterName = param.Identifier.ToString();
            if (!parameterName.StartsWith("@")) parameterName = $"@{parameterName}";
            if (identifierName != parameterName)
                return false;
            var mae = iName.Parent as MemberAccessExpressionSyntax;
            if (mae == null)
                return true;
            return mae.DescendantNodes().FirstOrDefault() == iName;
        }

        private static bool IsCandidateForRemoval(BaseMethodDeclarationSyntax methodOrConstructor, SemanticModel semanticModel)
        {
            if (methodOrConstructor.Modifiers.Any(m => m.ValueText == "partial" || m.ValueText == "override" || m.ValueText == "abstract" || m.ValueText == "extern")
                || !methodOrConstructor.ParameterList.Parameters.Any())
                return false;
            var method = methodOrConstructor as MethodDeclarationSyntax;
            if (method != null)
            {
                if (method.ExplicitInterfaceSpecifier != null) return false;
                var methodSymbol = semanticModel.GetDeclaredSymbol(method);
                if (methodSymbol == null) return false;
                var typeSymbol = methodSymbol.ContainingType;
                if (typeSymbol.AllInterfaces.SelectMany(i => i.GetMembers())
                    .Any(member => methodSymbol.Equals(typeSymbol.FindImplementationForInterfaceMember(member))))
                    return false;
                if (IsEventHandlerLike(method, semanticModel)) return false;
                if (IsPrivateAndUsedAsMethodGroup(method, methodSymbol, semanticModel)) return false;
                if (method.Parent is InterfaceDeclarationSyntax) return false;
            }
            else
            {
                var constructor = methodOrConstructor as ConstructorDeclarationSyntax;
                if (constructor != null)
                {
                    if (IsSerializationConstructor(constructor, semanticModel)) return false;
                }
                else
                {
                    return false;
                }
            }
            return true;
        }

        private static bool IsPrivateAndUsedAsMethodGroup(MethodDeclarationSyntax method, IMethodSymbol methodSymbol, SemanticModel semanticModel)
        {
            if (methodSymbol.DeclaredAccessibility != Accessibility.Private) return false;
            var parentType = method.Parent;
            var allTokens = parentType.DescendantTokens();
            var tokensThatMatch = from t in allTokens
                                  let text = t.Text
                                  where text == method.Identifier.Text
                                  select t;
            foreach (var token in tokensThatMatch)
            {
                var nodeSymbol = semanticModel.GetSymbolInfo(token.Parent).Symbol;
                if (methodSymbol.Equals(nodeSymbol)) return true;
            }
            return false;
        }

        private static bool IsSerializationConstructor(ConstructorDeclarationSyntax constructor, SemanticModel semanticModel)
        {
            if (constructor.ParameterList.Parameters.Count != 2) return false;
            var constructorSymbol = semanticModel.GetDeclaredSymbol(constructor);
            var typeSymbol = constructorSymbol?.ContainingType;
            if (!typeSymbol?.AllInterfaces.Any(i => i.ToString() == "System.Runtime.Serialization.ISerializable") ?? true) return false;
            if (!typeSymbol.GetAttributes().Any(a => a.AttributeClass.ToString() == "System.SerializableAttribute")) return false;
            var serializationInfoType = semanticModel.GetTypeInfo(constructor.ParameterList.Parameters[0].Type).Type as INamedTypeSymbol;
            if (serializationInfoType == null) return false;
            if (!serializationInfoType.AllBaseTypesAndSelf().Any(type => type.ToString() == "System.Runtime.Serialization.SerializationInfo"))
                return false;
            var streamingContextType = semanticModel.GetTypeInfo(constructor.ParameterList.Parameters[1].Type).Type as INamedTypeSymbol;
            if (streamingContextType == null) return false;
            return streamingContextType.AllBaseTypesAndSelf().Any(type => type.ToString() == "System.Runtime.Serialization.StreamingContext");
        }

        private static bool IsEventHandlerLike(MethodDeclarationSyntax method, SemanticModel semanticModel)
        {
            if (method.ParameterList.Parameters.Count != 2
                || method.ReturnType.ToString() != "void")
                return false;
            var senderType = semanticModel.GetTypeInfo(method.ParameterList.Parameters[0].Type).Type;
            if (senderType.SpecialType != SpecialType.System_Object) return false;
            var eventArgsType = semanticModel.GetTypeInfo(method.ParameterList.Parameters[1].Type).Type as INamedTypeSymbol;
            if (eventArgsType == null) return false;
            return eventArgsType.AllBaseTypesAndSelf().Any(type => type.ToString() == "System.EventArgs");
        }

        private static SyntaxNodeAnalysisContext ReportDiagnostic(SyntaxNodeAnalysisContext context, ParameterSyntax parameter)
        {
            var props = new Dictionary<string, string> { { "identifier", parameter.Identifier.Text } }.ToImmutableDictionary();
            var diagnostic = Diagnostic.Create(Rule, parameter.GetLocation(), props, parameter.Identifier.ValueText);
            context.ReportDiagnostic(diagnostic);
            return context;
        }
    }
}
